src/sqswatcher/sqswatcher.py

```diff
@@ -239,27 +239,28 @@ def _parse_sqs_messages(sqs_config_region, sqs_config_proxy, messages, table, qu
             continue
 
         if event_type == "parallelcluster:COMPUTE_READY":
-            update_event = _process_compute_ready_event(
-                sqs_config_region, sqs_config_proxy, message_attrs, message, table
-            )
+            add_event = _process_compute_ready_event(sqs_config_region, sqs_config_proxy, message_attrs, message, table)
+            if add_event:
+                add_events.append(add_event)
         elif event_type == "autoscaling:EC2_INSTANCE_TERMINATE":
-            update_event = _process_instance_terminate_event(message_attrs, message, table, queue)
+            remove_event = _process_instance_terminate_event(message_attrs, message, table, queue)
+            if remove_event:
+                remove_events.append(remove_event)
         else:
             log.info("Unsupported event type %s. Discarding message." % event_type)
-            update_event = None
-
-        if update_event:
-            log.info("Processing %s event for instance %s", event_type, update_event.host)
-            hostname = update_event.host.hostname
-            if hostname in update_events:
-                # delete first to preserve messages order in dict
-                del update_events[hostname]
-            update_events[hostname] = update_event
-        else:
-            # discarding message
-            log.warning("Discarding message %s", message)
             message.delete()
 
+    update_events = OrderedDict()
+    for event in itertools.chain(add_events, remove_events):
+        log.info("Processing %s event for instance %s", event.action, event.host)
+        hostname = event.host.hostname
+        if hostname in update_events:
+            # events are looped in the order they are fetched and by iterating over REMOVE events after ADD events.
+            # in case of collisions let's always remove the item that comes first in order to always favor REMOVE ops.
+            log.info("Hostname collision. Discarding event %s in favour of event %s", update_events[hostname], event)
+            del update_events[hostname]
```

Comment:
What happens to the message associated to the overwritten event? 
I don't see we're removing it from the queue. 
This flaw is not specifically of this patch but that a potential source of for large clusters.

Reply:
Good catch. colliding messages need to be removed from the queue.
In general to make this more robust we should filter messages based on the actual nodes available in ec2 so that we never process nodes that had been already terminated.

I added a quick fix for this: https://github.com/aws/aws-parallelcluster-node/pull/201/commits/210f6db95248aed59fa2a9828fce9dc113fb284a